moshi/src/main/java/com/squareup/moshi/JsonAdapter.java

```diff
@@ -26,12 +28,25 @@
 public abstract class JsonAdapter<T> {
   public abstract T fromJson(JsonReader reader) throws IOException;
 
+  public final T fromJson(Source source) throws IOException {
```

Comment:
Does this lose data?

```
Buffer b = new Buffer().writeUtf8("{}{}");
adapter.fromJson(b);
adapter.fromJson(b);
```

When we create a buffer, we risk reading more than we use.
.